azure-iot-device/azure/iot/device/provisioning/transport/state_based_mqtt_provider.py

```diff
@@ -0,0 +1,377 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import logging
+import six.moves.queue as queue
+import six.moves.urllib as urllib
+from transitions import Machine
+from . import constant
+from azure.iot.device.iothub.transport.mqtt.mqtt_provider import MQTTProvider
+from ..internal.action import SubscribeAction, PublishAction, UnsubscribeAction
+
+logger = logging.getLogger(__name__)
+
+
+class StateBasedMQTTProvider(object):
```

Comment:
>StateBasedMQTTProvider [](start = 6, length = 22)

This is awesome!  I have almost the exact same file in one of my branches, only I started from a different place (carter's provider refactor) and had a few minor differences (mostly about sas tokens, which I don't think belong at this level, but I don't think you should worry about that just yet).

Reply:
For what it's worth, in my version, the __init__ parameters are (self, client_id, hostname, username, ca_cert=None) since this is how the provider is initialized and the rules to define these values are very iothub (or dps) specific and I wanted this layer to be very MQTT specific with zero knowledge of IOTHub or DPS.  I also added a reconnect function (because that's what we do at the MQTT level for sas-token updates), and I put "password" as a parameter of connect and reconnect (since that's what the sas-token produces in the end) and that mirrors the provider function below.  (I'm still having some doubts about this).

---
In reply to: [276845842](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r276845842) [](ancestors = 276845842)

Reply:
yes makes sense...we can talk more and i can make changes accordingly to the request response in that case

---
In reply to: [276847616](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r276847616) [](ancestors = 276847616,276845842)

Reply:
resolving this as of now since I can not make changes without the changes to the transport itself.

---
In reply to: [276867201](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r276867201) [](ancestors = 276867201,276847616,276845842)